velox/common/caching/AsyncDataCache.cpp

```diff
@@ -503,5 +503,63 @@ std::string AsyncDataCache::toString() const {
       << stats.numPrefetch << " Alloc Mclks " << (stats.allocClocks >> 20);
   return out.str();
 }
+ReadPinsResult readPins(
+    const std::vector<CachePin>& pins,
+    int32_t maxGap,
+    std::function<uint64_t(const CachePin& pin, int32_t index)> offsetFunc,
+    std::function<void(
+        const std::vector<CachePin>& pins,
+        int32_t begin,
+        int32_t end,
+        uint64_t offset,
+        const std::vector<folly::Range<char*>>& buffers)> readFunc) {
+  std::vector<folly::Range<char*>> buffers;
+  uint64_t start = offsetFunc(pins[0], 0);
+  uint64_t lastOffset = start;
+  ReadPinsResult result;
+  int32_t firstPin = 0;
+  for (auto i = 0; i < pins.size(); ++i) {
+    auto& pin = pins[i];
+    VELOX_CHECK(pin.entry()->key().fileNum.hasValue());
+    auto& data = pin.entry()->data();
+    uint64_t startOffset = offsetFunc(pin, i);
+    result.readBytes += pin.entry()->size();
+    if (lastOffset < startOffset) {
+      auto gap = startOffset - lastOffset;
+      if (gap < maxGap) {
+        result.extraBytes += gap;
+        buffers.push_back(
+            folly::Range<char*>(nullptr, startOffset - lastOffset));
+      } else {
+        readFunc(pins, firstPin, i, start, buffers);
```

Comment:
The readFunc API might be simpler if we passes two CachePin iterators - start and end - and buffers. It seems redundant to pass 'start' offset as well.